### DISCUSSION TITLE
Getting rid of the core <> periphery terminology

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#1068] [Ideas]
Since the addition of `MerkleInstant` contract, I wonder if the core <> periphery still makes sense?

The `MerkleInstant` contract holds the airdrop funds, which are never deposited in any `core` contract.

What do you think, @sablier-labs/solidity?

**Comment by andreivladbrg** [comment]
Just to confirm, is this what you are suggesting?


From: 
<details><summary>Click to see tree</summary>
<p>

```tree
src/
├── core
│   ├── LockupNFTDescriptor.sol
│   ├── SablierLockupDynamic.sol
│   ├── SablierLockupLinear.sol
│   ├── SablierLockupTranched.sol
│   ├── abstracts
│   │   ├── Adminable.sol
│   │   ├── NoDelegateCall.sol
│   │   └── SablierLockup.sol
│   ├── interfaces
│   │   ├── IAdminable.sol
│   │   ├── ILockupNFTDescriptor.sol
│   │   ├── ISablierLockup.sol
│   │   ├── ISablierLockupDynamic.sol
│   │   ├── ISablierLockupLinear.sol
│   │   ├── ISablierLockupRecipient.sol
│   │   └── ISablierLockupTranched.sol
│   ├── libraries
│   │   ├── Errors.sol
│   │   ├── Helpers.sol
│   │   ├── NFTSVG.sol
│   │   └── SVGElements.sol
│   └── types
│       └── DataTypes.sol
└── periphery
    ├── SablierBatchLockup.sol
    ├── SablierMerkleFactory.sol
    ├── SablierMerkleInstant.sol
    ├── SablierMerkleLL.sol
    ├── SablierMerkleLT.sol
    ├── abstracts
    │   └── SablierMerkleBase.sol
    ├── interfaces
    │   ├── ISablierBatchLockup.sol
    │   ├── ISablierMerkleBase.sol
    │   ├── ISablierMerkleFactory.sol
    │   ├── ISablierMerkleInstant.sol
    │   ├── ISablierMerkleLL.sol
    │   └── ISablierMerkleLT.sol
    ├── libraries
    │   └── Errors.sol
    └── types
        └── DataTypes.sol
```

</p>
</details> 

To:

<details><summary>Click to see tree</summary>
<p>

```tree
src/
├── LockupNFTDescriptor.sol
├── SablierLockupDynamic.sol
├── SablierLockupLinear.sol
├── SablierLockupTranched.sol
├── SablierBatchLockup.sol
├── SablierMerkleFactory.sol
├── SablierMerkleInstant.sol
├── SablierMerkleLL.sol
├── SablierMerkleLT.sol
├── abstracts
│   ├── Adminable.sol
│   ├── NoDelegateCall.sol
│   ├── SablierLockup.sol
│   └── SablierMerkleBase.sol
├── interfaces
│   ├── IAdminable.sol
│   ├── ILockupNFTDescriptor.sol
│   ├── ISablierLockup.sol
│   ├── ISablierLockupDynamic.sol
│   ├── ISablierLockupLinear.sol
│   ├── ISablierLockupRecipient.sol
│   ├── ISablierLockupTranched.sol
│   ├── ISablierBatchLockup.sol
│   ├── ISablierMerkleBase.sol
│   ├── ISablierMerkleFactory.sol
│   ├── ISablierMerkleInstant.sol
│   ├── ISablierMerkleLL.sol
│   └── ISablierMerkleLT.sol
├── libraries
│   ├── Errors.sol
│   ├── Helpers.sol
│   ├── NFTSVG.sol
│   └── SVGElements.sol
└── types
    └── DataTypes.sol
```

</p>
</details> 

Will there be a common `DataTypes.sol` and `Errors.sol` files for both?

**Reply by PaulRBerg** [reply]
Yes and yes.

**Comment by smol-ninja** [comment]
Interesting. If you are specifically targeting `MerkleInstant` contract then in its defence, I can argue that its similar to `MerkleLL` and `MerkleLT`. They also hold airdrop funds until user "claims" them.

To @andreivladbrg, an argument against combining core <> periphery into one is that the merkle lockup contracts do not form the lockup protocol. The current periphery is made of two parts:

1. **Batch contract**: batches `create` functions of lockup protocol. Since, its tightly integrated with the Lockup's create APIs, I wonder why it should not be a part of the core protocol itself. The core protocol also contains batch functions such as `withdrawMultiple`, `withdrawMaxAndTransfer`, and now will include the `batch` function, so why give batch create a special attention.


2. **Airstreams**: They are not a part of the Lockup protocol. They have their own system of contracts which can be used to facilitate airdrops either through lockups or instant or some other upcoming system.

Having said that, "Airstreams" protocol deserve its own repo. But if we are interested into keeping them in Lockup repo, a better structure would be the following:
1. Include `BatchLockup` in the core directory
2. Rename the periphery to airstreams

**Reply by andreivladbrg** [reply]
> They have their own system of contracts which can be used to facilitate airdrops either through lockups or instant or some other upcoming system

the majority (2 vs 1) are using the lockup contracts, so they are related

> Having said that, "Airstreams" protocol deserve its own repo. 

hmm, isn't this in contradictory to package tethering?

> Include BatchLockup in the core directory
Rename the periphery to airstreams

that's a good idea, i like it. also, we could rename `core` to something else, but i don't have a good name in mind

**Reply by smol-ninja** [reply]
> isn't this in contradictory to package tethering

Good question. Yes it is contradictory to package tethering. Lets forget this idea.

> the majority (2 vs 1) are using the lockup contracts, so they are related

Technically yes they are related. If we make change to create APIs, we would have to change them as well so yes they are related.

> that's a good idea, i like it. also, we could rename `core` to something else, but i don't have a good name in mind

Fantastic. From the high level, we can read them as `lockup/core` and `lockup/airstreams`.

**Reply by PaulRBerg** [reply]
To @smol-ninja:

- `MerkleLL` and `MerkleLT` interact with the core contracts. `MerkleInstant` doesn't. This is a qualitative difference.
- I agree with moving the batch contract to the `core`. But, while we are at it ... why not implement `createMultiple` functions in the `Lockup` contract directly? Contract size limits?
- Interesting idea to introduce a new repo for Airstreams. I'd call it Airdrops, though. My only reservation is that Airdrops interacts with Lockup, which is not the case between Lockup and Flow.
- I don't like the idea of using `lockup/core` and `lockup/airstreams`. It feels a bit off — the first is the core of Lockup, and the second is ... something else? The categories are not at the same logical depth.

To both:

- Introducing a new repo for Airstreams wouldn't break package tethering. Removing the `MerkleLockup` would be a breaking change, but the package tethering is in itself a breaking change (moving from `@sablier/v2-core` to `@sablier/lockup`), so we're all good.

**Reply by PaulRBerg** [reply]
Let me sleep more on this proposal, but I am tentatively in favor of creating a separate repo for the Merkle airdrop contracts.

**Reply by smol-ninja** [reply]
> why not implement createMultiple functions in the Lockup contract directly?

Yes, the contract size limitations. I had to remove three functions to [fit everything into a single contract](https://github.com/sablier-labs/v2-core/discussions/1064#discussioncomment-11026956). If achieving a singleton isn’t a priority, we can consider including them..

> My only reservation is that Airdrops interacts with Lockup

Yes, but this only applies to 66% of them. In the future, we might also consider the possibility of developing more airdrop contracts that do not rely on lockups (such as [this](https://github.com/sablier-labs/company-discussions/discussions/85)). So having a separate repo for Airstreams keep that option open for us.

> feels a bit off — the first is the core of Lockup, and the second is ... something else

I agree.

> Introducing a new repo for Airstreams wouldn't break package tethering

That sounds reasonable. Airstream contracts could also have their own versioning system.

----------------

Instead of organizing them as a monorepo, it would then be restructured as "product oriented repos": lockup, flow and airstream (name tbd).

Here is [package tethering discussion](https://github.com/sablier-labs/company-discussions/discussions/25) to remind ourselves of all the arguments for monorepo design just in case.

**Reply by PaulRBerg** [reply]
These are all great responses.

I am open to setting up the separate repo for airdrops.

**Reply by smol-ninja** [reply]
I am also in favour of that. cc @andreivladbrg whats your view on this?

**Reply by andreivladbrg** [reply]
#### re airstream dedicated repo

assuming that this change is going to be post `1.3.0`, the a problem i see with this is the overload of packages with different versioning. our app (or maybe others that have integrated us) would have a problem with backwards compatibility: `@sablier-labs/v2-core#versionx.y.z && @sablier-labs/v2-periphery#versionx.y.z && @sablier-labs/lockup#versionx.y.z && @sablier-labs/airstreams`.
cc @razgraf RFF

also, another argument against this (also mentioned in the package tethering discussion) is that expanding to other chains would require having 2 different repos per chain, which might not be as easy to handle

and lastly, i believe, since the name of the repo is going to be “lockup”, it could be suitable to have the airstream campaign that does not use the streaming contract (LL/LT) in this repo, as the campaign owner will be locking up some funds in the campaign, right?

nevertheless, i am tentatively against this for now, until we reach a number of campaigns that do not use a streaming contract greater than the ones that use them

**Reply by PaulRBerg** [reply]
Re what you said in the first first paragraph — addressed it in my [comment above](https://github.com/sablier-labs/v2-core/discussions/1068#discussioncomment-11031804) when I said package tethering is a breaking change.

Re expansion to other chains — it will take time. The first version on Solana is just a Lockup implementation. But also, there's nothing wrong with having multiple repos for different blockchain paradigms. In fact, it may be beneficial.

> it could be suitable to have the airstream campaign that does not use the streaming contract (LL/LT) in this repo

I disagree based on the explanation below.

> as the campaign owner will be locking up some funds in the campaign

The Lockup refers to the recipient's perspective.

With instant airdrops, there's no lockup for recipients. They claim right away.

**Reply by andreivladbrg** [reply]
> Re what you said in the first first paragraph — addressed it in my [comment above](https://github.com/sablier-labs/v2-core/discussions/1068#discussioncomment-11031804) when I said package tethering is a breaking change.

And I guess my point is whether this breaking change is truly useful, as it might impact more components than just where the smart contracts are placed. I know it’s impossible to predict everything in advance, but would the total development work vs. the improvement gained be worth it in the short/long term?

> it will take time

Given enough time, everything is possible. It is about prioritization I guess🤔

> But also, there's nothing wrong with having multiple repos for different blockchain paradigms.

Absolutely, but I think the question isn’t about different repos for different blockchains (which I agree, they should be different), but rather whether _streaming_ and _airstream_ contracts should be in the same repo or not (since most of them use the streaming contract).

We could either have one repo per blockchain (lockup – including both streaming and airstream contracts) or two repos per blockchain. 

`1. evm-lockup 2. sol-lockup` **vs** `1. evm-lockup 2. evm-airstreams 3. sol-lockup 4. sol-airstreams` - (don't mind the repo names)

> The Lockup refers to the recipient's perspective.
With instant airdrops, there's no lockup for recipients. They claim right away.

You’re correct. However, in defense of the terminology, at some point, an actor will need to “lock up” some funds.

**Reply by PaulRBerg** [reply]
> whether this breaking change is truly useful

Yes. Otherwise, we wouldn't have implemented package tethering.

My point was that tethering the package is, in itself, a breaking change. Separating the airdrop contracts wouldn’t alter this from a qualitative perspective.

Do you see this now?

> rather whether streaming and airstream contracts should be in the same repo or not (since most of them use the streaming contract).

Yes, I understood that. What I meant to say is that it'll be fine to have a separate airdrop repo for the contracts on each chain.

> two repos per blockchain.

Yes.

> would the total development work vs. the improvement gained be worth it in the short/long term

Yes as per the arguments given by Shub and I throughout this discussion.

> an actor will need to “lock up” some funds

Yes. Words have multiple meanings. We should use and promote the meaning that makes most sense.

**Reply by andreivladbrg** [reply]
>My point was that tethering the package is, in itself, a breaking change. Separating the airdrop contracts wouldn’t alter this from a qualitative perspective.
Do you see this now?

i see, but i have mentioned [this](https://github.com/sablier-labs/v2-core/discussions/1068#discussioncomment-11073753) in my comment from the beginning: "assuming that this change is going to be post 1.3.0"  - post the current package tethering version, which implies we would have two breaking changes consecutively (about this, the question about "improvement gained be worth it in the short/long term" is raised)

> Yes as per the arguments given by Shub and I throughout this discussion.

yes, i assume you’re referring to these two points:

>>Yes, but this only applies to 66% of them. In the future, we might also consider the possibility of developing more airdrop contracts that do not rely on lockups (such as https://github.com/sablier-labs/company-discussions/discussions/85). So having a separate repo for Airstreams keep that option open for us.
>> Instead of organizing them as a monorepo, it would then be restructured as "product oriented repos": lockup, flow and airstream

and my response was:

>> nevertheless, i am tentatively against this for now, until we reach a number of campaigns that do not use a streaming contract greater than the ones that use them

which wasn’t answered

**Reply by PaulRBerg** [reply]
Oh, sorry I missed the part about post-1.3.0 — the thing is, however, that this change is indeed to be included in v1.3.0

> i am tentatively against this for now, until we reach a number of campaigns that do not use a streaming contract greater than the ones that use them

The fact that some contracts are no longer interacting with Lockup is good enough of a reason to create the separate repo.

Categories are useful even when there are no clear, sharp distinction between them.

So, to go back to the problem in the OP — we have a problem of category formation.

The periphery is no longer a periphery because some contracts are completely independent of the core contracts.

By moving the airdrop contracts to a separate repo, we would get rid of this category problem.

It's fine to put the `MerkleLL` & co. contracts in the separate repo because they are airdrop contracts. It's just a more natural place for them — the fact that they are airdrop contracts is a more defining characteristic than the fact that they interact with the core contracts.

To give a further example: we also use OpenZeppelin's contracts, but we don't open PRs to move our code there.

**Reply by andreivladbrg** [reply]
> Oh, sorry I missed the part about post-1.3.0 — the thing is, however, that this change is indeed to be included in v1.3.0

Great, then we’re having a different discussion.

I’m happy to say I’ve changed my mind and agree with this. Thanks for re-explaining

**Reply by PaulRBerg** [reply]
Cheers

**Comment by smol-ninja** [comment]
So, what do we call our new baby repo? Airstreams? Merkle-Drop? Any ideas?

**Reply by PaulRBerg** [reply]
`airdrops`

**Reply by PaulRBerg** [reply]
Or maybe `merkle`?

Just like the `lockup` repo is named after the name of the contracts (`Lockup`).

**Reply by smol-ninja** [reply]
`merkle` is popularly known to represents merkle tree. So the name can be confused with merkle libraries. `airdrops` is a very generic yet good name. However, the names can be confused with "Sablier doing an airdrop". I guess your argument against `airstreams` is that this will no longer facilitate only streams. So I agree with you on that.

**Reply by PaulRBerg** [reply]
> the names can be confused with "Sablier doing an airdrop"

Not bad for marketing

**Reply by smol-ninja** [reply]
Should we rename `v2-periphery` to `airdrops` instead of creating a new repo?

**Reply by PaulRBerg** [reply]
[No](https://github.com/sablier-labs/v2-core/issues/1072#issuecomment-2478088347).

**Comment by smol-ninja** [comment]
### Issues
1. https://github.com/sablier-labs/v2-core/issues/1072
2. https://github.com/sablier-labs/v2-core/issues/1073
